### PR TITLE
Allow running in debug product mode with checked mode off.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Dart code runs in checked mode when the runtime mode is debug. In profile and release product modes, the application code is precompiled and checked mode is unsupported. However, this flag may be specified if the user wishes to run in the debug product mode (i.e. with JIT or DBC) with checked mode off.

I remember we had this flag earlier. Not sure when we removed the same.

### CODE REVIEW

**Comment by jason-simmons** [comment]
LGTM

**Comment by eseidelGoogle** [comment]
I think we did it as part of simplifying what modes we have:
https://github.com/flutter/flutter/wiki/Flutter%27s-modes
@Hixie @abarth

**Comment by chinmaygarde** [comment]
> What's the use case for this mode?

Disabling checked mode in debug. Since the desktop and simulator shells are never in profile and release modes, it is impossible to disable checked mode on the desktop or simulator. I usually just had a patch stashed that commented out adding the flags while I tinkered on the engine (extremely stuttery otherwise). Turns our @mishadynin and company wanted to do the same. Hence the flag. Should not change workflow for external developers or anything in the tooling. But you should be able to add the flag via Xcode (for the simulator) or via the command line (for the desktop shell).

**Comment by Hixie** [comment]
I meant, what's the use case for disabling checked mode in debug?

**Comment by chinmaygarde** [comment]
Checked mode makes the application slow and stuttery. Developers who are working on the engine or otherwise don't care about assertions on the Dart frameworks side can use use this flag for a smoother experience.

**Comment by Hixie** [comment]
If we care about performance on simulators then we should do a proper job and have a fast experience on simulators, for example doing proper release builds, or adding a new officially-supported mode, or whatever it takes.

We shouldn't have team-specific workarounds that hide the pain for us if our users are still going to feel the pain. If the pain is bad enough for us to work around then it is definitely bad enough for us to fix it for our users. We also shouldn't have modes that aren't regularly and properly tested. Such modes are very likely to be a source of bugs, potentially security bugs, not to mention increasing the complexity of the product.

**Comment by abarth** [comment]
IMHO, we should revert this patch.  I don't think we should expand the matrix of supported configurations.  If debug mode is too slow, we should profile it and make it faster.

**Comment by chinmaygarde** [comment]
This patch improves the quality of life of developers working on the engine. It is in no way intended to hide or suppress the supposed pain of our users who develop in debug mode on the simulators and devices.

Think of it this way; we optimize native code, disable assertions, remove verbose logging, etc. from the engine in all modes for devices (including the debug mode). We reason that these assertions, while useful to developers working on the engine, are of limited use to the users who interact with the framework. More importantly, if we were to keep the code unoptimized (at -O0), it would not reflect the true capabilities of the engine and its dependencies. However, our users do care deeply about all the assertions in the Dart framework. These assertions give helpful warnings about potential incorrect use of the framework.

Now all you need to do is to flip roles. Engine developers usually compile the engine as unoptimized (so debuggers can give sane backtraces) with full logging and assertions. While doing this, they don’t care as much about all the optional developer aids the framework provides (and slows itself down to service). All this patch does is to give them the ability to opt of the stuff they don’t care about so they can have a better experience. I know I feel proud of the product I work on when the scrolling is buttery smooth as I tinker on the engine.

The implication that anything in this patch distracts us from our responsibility of doing what is best for our users (by shrugging real problems under the proverbial carpet) is slightly irksome. Especially since all developer aids are written to make servicing the needs of our users easier. 

Also, this patch in no way affects our product vision. We have freely added and removed flags from the engine without it affecting the Flutter tools interface. Our users don’t interact with the engine via flags, they do so via the Flutter tools and the embedder APIs. My hope is that all command line flags will be disable when the application is released.

About the support matrix; we have been aggressive in our stance that we only support the two mobile platforms. Nothing in this patch changes that. In fact, even the `--interactive` flag that is the foundation of my developer workflow is an unsupported configuration. I realize that. I maintain it because it makes my life easier. In fact, I think the last time someone used that flag on Linux was John when he was working the hot reload (many months ago). So I am not even sure support exists for the flag on Linux any more. This configuration is not in our support matrix. We have never implied this.

Bottom line, this patch makes the lives of engine developers easier and affects neither the product vision nor how we communicate that vision to our users. For this reason, we should keep this patch, and, continue adding patches that allow us to be nimble in making the best product for our users.

**Comment by abarth** [comment]
Ok, it sounds you feel strongly about this topic.  We can keep the flag.

I do think we're papering over checked mode being a bit too slow.  I've noticed the same thing recently when working with the gallery.  However, we can work on that problem separately.

**Comment by Hixie** [comment]
I'm still concerned that just turning off assertions is enough for us to go from an unusable experience to a buttery-smooth experience. Can you quantify this more? What app are you running, what actual difference does it make, etc?

(FWIW, I usually run things with locally built unopt builds, even on devices.)

**Comment by chinmaygarde** [comment]
The script I run locally eventually boils down to a call like the one I pasted below. This one is for the gallery but really anything that has scrolling or has a drawer stutters quite a bit. Try the invocation once without the new flag.
```
lldb -- ./out/host_debug_unopt/SkyShell.app/Contents/MacOS/SkyShell  \
  --flx=path_to_flutter/examples/flutter_gallery/build/app.flx       \
  --dart-main=path_to_flutter/examples/flutter_gallery/lib/main.dart \
  --packages=path_to_flutter/examples/flutter_gallery/.packages      \
  --observatory-port=8181                                            \
  --trace-startup                                                    \
  --dart-non-checked-mode                                            \
```
To be clear, I don't mind that the checked mode is slow. It is doing a ton of useful checking that I know I would need when I am working on developing the application. But when I am not working an application, I'd rather turn it off using a flag instead of commenting out C++ code in the engine.

**Comment by Hixie** [comment]
Is it slow on devices as well, or is this specific to the host_debug_unopt mode?
